Support torchbind in OSS proxy executor

yushangdi opened the pull request:
Summary:
Implement torchbind support in OSSProxyExecutor.

Exactly the same as the implementation in FbProxyExecutor.

D69693697 - fbProxyExecutor
D69887230 - fbProxyExecutor but for torchbind method


Other changes:

- When generating the schema of the CallTrochBind HOP, the arg name of the torchbind object arg should be the same as the torchbind method's torchbind object arg (instead of `obj`).

- In `AOTIModelPackageLoader`, we extract everything in `data/constants` to `tmp_dir/data/aot_inductor/<model>/` folder, so the torchbind objs exist in the same folder as the rest of the files (e.g. cpp, so). This is to be consistent of how files are packaged internally

Test Plan:
```
buck run fbcode//mode/dev-nosan //caffe2/test/inductor:torchbind -- -r torchbind_aoti

buck run fbcode//mode/dev-nosan //caffe2/test/inductor:torchbind -- -r aot_compile
```

Differential Revision: D69500038




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Skylion007 commented on torch/csrc/inductor/aoti_package/model_package_loader.cpp:
```diff
@@ -63,6 +63,11 @@ const std::string k_separator = "\\";
 const std::string k_separator = "/";
 #endif
 
+bool startsWith(const std::string& str, const std::string& prefix) {
```
C10 has a shim for this already.

Skylion007 replied:
Or better yet: https://github.com/pytorch/pytorch/blob/main/c10/util/string_view.h

Skylion007 commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -47,13 +48,26 @@ struct OSSDynamicArg {
 
+struct OSSTorchBindArg {
+  OSSTorchBindArg(int arg_index, std::string arg_name)
+      : arg_index(arg_index), arg_name(arg_name) {}
```
Arg name should be moved into the initializer std::move

zou3519 reviewed (commented):
_higher_order_ops change LGTM, someone else who knows the AOTI code should take a look at that

desertfire commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:
```diff
@@ -18,20 +32,46 @@ void OSSProxyExecutor::prefill_stack_with_static_arguments(
-    OSSOpKernel& op_kernel) {
-  auto& stack = op_kernel.stack_;
-  auto& dynamic_args = op_kernel.dynamic_args_;
+    OSSOpKernel* op_kernel,
```
Curious why this refernce to pointer change?

desertfire replied:
Never mind. I see your created child classes later.

desertfire commented on torch/csrc/inductor/aoti_package/model_package_loader.cpp:
```diff
@@ -389,11 +390,27 @@ AOTIModelPackageLoader::AOTIModelPackageLoader(
+      } else { // startsWith(filename_str, const_directory)
+        // Extract constants to the same directory as the rest of the files
+        // to be consistent with internal implementation
+        size_t lastSlash = filename_str.find_last_of("/\\");
```
You can just use k_separator here.

yushangdi replied:
> You can just use k_separator here.

changed to use `k_separator` now

yushangdi committed:
Support torchbind in OSS proxy executor (#149747)

Summary:

Implement torchbind support in OSSProxyExecutor.

Exactly the same as the implementation in FbProxyExecutor.

D69693697 - fbProxyExecutor
D69887230 - fbProxyExecutor but for torchbind method
D70746626 - Support None output type

Other changes:

- When generating the schema of the CallTrochBind HOP, the arg name of the torchbind object arg should be the same as the torchbind method's torchbind object arg (instead of `obj`).

- In `AOTIModelPackageLoader`, we extract everything in `data/constants` to `tmp_dir/data/aot_inductor/<model>/` folder, so the torchbind objs exist in the same folder as the rest of the files (e.g. cpp, so). This is to be consistent of how files are packaged internally (more details in internal Diff summary).


Note on using `filesystem`:

Seems like there'll be [issues](https://github.com/pytorch/pytorch/pull/137209) with using`filesystem` header in linux, so here I use string manipulation instead of `filesystem::path`.

Test Plan:
```
buck run fbcode//mode/dev-nosan //caffe2/test/inductor:torchbind -- -r torchbind_aoti

buck run fbcode//mode/dev-nosan //caffe2/test/inductor:torchbind -- -r aot_compile

buck2 build --flagfile fbsource//arvr/mode/win/dev fbsource//arvr/libraries/audio/AudioSDK/Research/Source/BSI/UnitTests:meridian_bsi_tests

buck2 build --flagfile fbsource//arvr/mode/embedded/linux/clang-aarch64-release fbsource//arvr/third-party/signalarity/latest-release/src/shaker/ha:signalarity_realtime_libtorch_pipeline_
```

Reviewed By: desertfire

Differential Revision: D69500038

yangw-dev commented:
@pytorchbot merge -i

yushangdi commented:
relanding in https://github.com/pytorch/pytorch/pull/150196
